lib/widgets/message_list.dart

```diff
@@ -523,11 +532,58 @@ class MarkAsReadWidget extends StatelessWidget {
                     height: (23 / 18))
                   .merge(weightVariableTextStyle(context, wght: 400))),
                 shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(7)),
+              ).copyWith(
+                // clobber `FilledButton`'s default disabled state
+                // TODO(#95) need dark-theme colors (foreground and background)
+                foregroundColor: WidgetStateColor.resolveWith((_) => Colors.white),
+                backgroundColor: WidgetStateColor.resolveWith((_) => _UnreadMarker.color),
               ),
-              onPressed: () => _handlePress(context),
+              onPressed: loading ? null : () => _handlePress(context),
               icon: const Icon(Icons.playlist_add_check),
-              label: Text(zulipLocalizations.markAllAsReadLabel))))),
-    );
+              label: Text(zulipLocalizations.markAllAsReadLabel))))));
+  }
+}
+
+class MarkAsReadAnimation extends StatefulWidget {
+  final bool loading;
+  final bool hidden;
+  final Widget child;
+
+  const MarkAsReadAnimation({
+    super.key,
+    required this.loading,
+    required this.hidden,
+    required this.child
+  });
+
+  @override
+  State<MarkAsReadAnimation> createState() => _MarkAsReadAnimationState();
+}
+
+class _MarkAsReadAnimationState extends State<MarkAsReadAnimation> {
+  bool _isPressed = false;
+
+  void _setScale(bool isPressed) {
+    setState(() {
+      _isPressed = isPressed;
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return GestureDetector(
+      onTapDown: (_) => _setScale(true),
+      onTapUp: (_) => _setScale(false),
+      onTapCancel: () => _setScale(false),
```

Comment:
Hmm, I don't love that we're effectively winding up here with a button that has two separate gesture detectors watching for a press — this one tracking whether it's currently being pressed, and the one the Material library supplies which we're using for responding to having been pressed.

In particular this feels like it's repeating the work that the Material button is already doing, and we could get a similar effect here by tracking the button's WidgetState, looking for WidgetState.pressed.

That's not a blocker for merging this, but it's something I'll want to clean up later if we end up sticking with this UI design.

Reply:
I agree! I've tried various approaches before, but I haven't found a straightforward solution that works as a replacement.

I'd love to hear your suggestions and I'd love to implement them.